common/locks/id_mutex.go

```diff
@@ -29,98 +29,38 @@ import (
 )
 
 type (
-	// HashFunc represents a hash function for string
-	HashFunc func(interface{}) uint32
+	// UnlockFunc unlocks a lock
+	UnlockFunc func()
 
 	// IDMutex is an interface which can lock on specific comparable identifier
 	IDMutex interface {
-		LockID(identifier interface{})
-		UnlockID(identifier interface{})
+		LockID(identifier interface{}) UnlockFunc
 	}
```

Comment:
this change makes IDMutex quite different from normal mutex?
seems that new change requires the caller to pass on the unlock function until where the lock is to be released

any reason to make this change?

Reply:
I originally made it because the process of looking up the right mutex was rather complicated: hash the identifier, index the slice of shards, lock the shard mutex, do a map lookup. We can just bypass all that work a second time by returning the unlock directly. With the simplification, it's just the lock and hash lookup, so not as big of a deal. I'd be open to changing it.

I would expect most or all uses of LockID to unlock in the same scope, so the unlock function can just sit in a local variable. How often do you think someone will need to unlock from a different function than they locked?

Reply:
depends on the use case, since this id mutex is currently not used & will only used by this PR, this change is fine to me.